precompiles/pointer/pointer.go

```diff
@@ -0,0 +1,257 @@
+		return nil, fmt.Errorf("error loading the pointer ABI %s", err)
+	}
+
+	newAbi, err := ethabi.JSON(bytes.NewReader(abiBz))
+	if err != nil {
+		return nil, err
+	}
+
+	p := &Precompile{
+		Precompile:  pcommon.Precompile{ABI: newAbi},
+		evmKeeper:   evmKeeper,
+		bankKeeper:  bankKeeper,
+		wasmdKeeper: wasmdKeeper,
+		address:     common.HexToAddress(PointerAddress),
+	}
+
+	for name, m := range newAbi.Methods {
+		switch name {
+		case AddNativePointer:
+			p.AddNativePointerID = m.ID
+		case AddCW20Pointer:
+			p.AddCW20PointerID = m.ID
+		case AddCW721Pointer:
+			p.AddCW721PointerID = m.ID
+		}
+	}
+
+	return p, nil
+}
+
+// RequiredGas returns the required bare minimum gas to execute the precompile.
+func (p Precompile) RequiredGas(input []byte) uint64 {
+	// gas is calculated dynamically
+	return 0
+}
+
+func (p Precompile) Address() common.Address {
+	return p.address
+}
+
+func (p Precompile) RunAndCalculateGas(evm *vm.EVM, caller common.Address, _ common.Address, input []byte, suppliedGas uint64, value *big.Int) (ret []byte, remainingGas uint64, err error) {
+	ctx, method, args, err := p.Prepare(evm, input)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	switch method.Name {
+	case AddNativePointer:
+		return p.AddNative(ctx, method, caller, args, value, evm, suppliedGas)
+	case AddCW20Pointer:
+		return p.AddCW20(ctx, method, caller, args, value, evm, suppliedGas)
+	case AddCW721Pointer:
+		return p.AddCW721(ctx, method, caller, args, value, evm, suppliedGas)
+	}
+	return
+}
+
+func (p Precompile) Run(*vm.EVM, common.Address, []byte, *big.Int) ([]byte, error) {
+	panic("static gas Run is not implemented for dynamic gas precompile")
+}
+
+func (p Precompile) AddNative(ctx sdk.Context, method *ethabi.Method, caller common.Address, args []interface{}, value *big.Int, evm *vm.EVM, suppliedGas uint64) (ret []byte, remainingGas uint64, err error) {
+	pcommon.AssertArgsLength(args, 1)
+	token := args[0].(string)
+	existingAddr, existingVersion, exists := p.evmKeeper.GetERC20NativePointer(ctx, token)
+	if exists && existingVersion >= native.CurrentVersion {
+		return nil, 0, fmt.Errorf("pointer at %s with version %d exists when trying to set pointer for version %d", existingAddr.Hex(), existingVersion, native.CurrentVersion)
+	}
+	metadata, metadataExists := p.bankKeeper.GetDenomMetaData(ctx, token)
+	if !metadataExists {
+		return nil, 0, fmt.Errorf("denom %s does not have metadata stored and thus can only have its pointer set through gov proposal", token)
+	}
+	name := metadata.Name
+	symbol := metadata.Symbol
+	decimals := uint8(0)
+	for _, denomUnit := range metadata.DenomUnits {
+		if denomUnit.Exponent > uint32(decimals) && denomUnit.Exponent <= math.MaxUint8 {
+			decimals = uint8(denomUnit.Exponent)
+			name = denomUnit.Denom
+			symbol = denomUnit.Denom
+			if len(denomUnit.Aliases) > 0 {
+				name = denomUnit.Aliases[0]
+			}
+		}
+	}
+	abi := native.GetABI()
+	parsedABI, err := ethabi.JSON(strings.NewReader(string(abi)))
```

Comment:
side thought: i wonder if it makes sense to cache the parsedABI in memory after first access?  (json memory/cpu can be expensive under load)

Reply:
good point - updated